### PR TITLE
issue/2824-react Added react template support

### PR DESCRIPTION
**Pull request by oliverfoster**
#2824 #2944 

Added react support and handlebars-style react templates.

These plugins achieve react templating and integration in the same way we have been doing with handlebars:
* [babel-plugin-transform-react-templates](https://github.com/cgkineo/babel-plugin-transform-react-templates/) (custom made) register templates from filenames
* [html-react-parser](https://www.npmjs.com/package/html-react-parser) parse html into react nodes, like triples braces in handlebars
* [plugin-transform-react-jsx](https://www.npmjs.com/package/@babel/plugin-transform-react-jsx) jsx compiler
* [react](https://www.npmjs.com/package/react) client-side react node handler
* [react-dom](https://www.npmjs.com/package/react-dom) client-side react node renderer

#### Templates
Supports react 'templates' as near as I could get it to our current handlebars templates - separate, override-able etc. This change does not require a wholesale switch to React views but instead uses our current Backbone views and React as a templating engine. This means that some of the usual functionality of React views, such as the state object, aren't used to cause view re-render, instead the `this.changed()` function should be called to re-render a Backbone View with a React template. 

Here are a few reasons for using React as a templating engine only:
1. Our entire architecture uses Backbone events heavily - which React views don't support. 
2. Many of our bugs arise from jQuery DOM manipulation and using React to manage the DOM manipulation removes those bugs whilst introducing much more flexibility in the view <> state updates, this is as each manipulation is now automated and no longer needs to be coded by hand. 
3. Introducting jsx as a templating engine allows the upskilling requirement to be confined to templates only, which should be easier and less daunting for developers to confront. 
4. We will be able to convert to React at our leisure by not mandating a wholesale view layer rewrite. 
5. As React will be a part of the architecture we can better experiment with dropping Backbone as a large part of the architectural work is already complete.

#### Usage

A React version of the accordion lives here [pr](https://github.com/adaptlearning/adapt-contrib-accordion/pull/97) / [branch](https://github.com/adaptlearning/adapt-contrib-accordion/tree/issue/2824-jsx)

```js
// This is an example template file in **/templates/**/name.jsx
// It will get registered in Adapt by filename as handlebars templates currently do
// Templates will override from theme, menu, extension, component as usual
// There is no partials distinction with the react templates, they are all just templates which can be called from one another

export default function (model, view) {
  return <div></div> // React component
};
```

```js
// This is a template file in components/adapt-contrib-accordion/templates/accordion.jsx

// These functions help handlebars integration
import Adapt from 'core/js/adapt';
import { 
  compile, // Compile handlebars into html: compile(template, data);
  classes, // Helper for joining and filtering classes
  templates, // Lookup other react templates: templates[name](model, view);
  html, // Render html as react html:  html(`<div>new div</div>`); or html(compile('{{#if _prop}}<div>new div</div>{{/if}'));
  partial // Use a handlebars partial: partial(name, data);
} from 'core/js/reactHelpers';

export default function (model, view) {
  const data = model.toJSON();
  data._globals = Adapt.course.get('_globals');
  return <div className="component__inner accordion__inner">
    {templates.component(model, view)}
    <div className="component__widget accordion__widget">
      {data._items.map(({ _graphic, _classes, title, body, _index, _isVisited, _isActive }, index) => {
        return <div key={_index} className={classes([
              "accordion__item",
              _graphic.src && 'has-image',
              _classes && _classes
          ])} data-index={_index}>
          <button onClick={view.onClick.bind(view)} id={`${data._id}-${index}-accordion-button`} className={classes([
            'accordion__item-btn',
            'js-toggle-item',
            _isVisited && 'is-visited',
            _isActive ? 'is-open is-selected' : 'is-closed'
          ])} aria-expanded={_isActive ? 'true' : 'false'}>
            <div className="accordion__item-btn-inner">
              <div className="accordion__item-icon">
                <div className="icon"></div>
              </div>
              <div className="accordion__item-title">
                <div className="accordion__item-title-inner">
                  {html(title)}
                </div>
              </div>
            </div>
          </button>
          <div className="accordion__item-content" role="region" aria-labelledby={`${data._id}-${index}-accordion-button`}>
            <div className="accordion__item-content-inner">
              {body &&
              <div className="accordion__item-body">
                <div className="accordion__item-body-inner">
                  {html(compile(body))}
                </div>
              </div>
              }
              {_graphic.src &&
              <div className={classes([
                  'accordion__item-image-container',
                  _graphic.attribution && 'has-attribution'
              ])}>
                <img className="accordion__item-image" src={_graphic.src} aria-label={_graphic.alt} aria-hidden={_graphic.alt ? 'true' : 'false'} />
                {_graphic.attribution &&
                <div className="component__attribution accordion__attribution">
                  <div className="component__attribution-inner accordion__attribution-inner">
                    {html(_graphic.attribution)}
                  </div>
                </div>
                }
              </div>
              }
            </div>
          </div>
        </div>
      })}
    </div>
  </div>
}
```

```js
// Example from accordion.
// adaptView is extended to support react templates, so all Adapt views can use react template as below.

import ComponentView from 'core/js/views/componentView';

class AccordionView extends ComponentView {

  preRender() {
    this.listenTo(this.model.getChildren(), {
      'all': this.changed // call changed to rerender
    });
  }

  /** missing code */

}

AccordionView.template = 'accordion.jsx'; // Just update the template name to use the react template

export default AccordionView;
```

### CODE REVIEW

File: src/core/templates/partials/component.jsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      <div className={classes(['component__header-inner', `${type}__header-inner`])}>
+        {displayTitle &&
+        <div className={classes(['component__title', `${type}__title`])}>
+
+          {!_disableAccessibilityState &&
+          <div className="js-heading" ref={view.jsxHeading}></div>
+          }
+
+          <div className={classes(['component__title-inner', `${type}__title-inner`])} aria-hidden={!_disableAccessibilityState}>
+            {`\n${html(compile(displayTitle, data))}\n`}
```
Comment (tomgreenfield) [inline src/core/templates/partials/component.jsx:33]: What's the reasoning behind the manual newline characters?

File: src/core/templates/partials/component.jsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    instruction,
+    mobileInstruction,
+    _component,
+    _disableAccessibilityState
+  } = data;
+  const type = _component.toLowerCase();
+  const sizedInstruction = (mobileInstruction && Adapt.device.screenSize !== 'large') ?
+    mobileInstruction :
+    instruction;
+  return (displayTitle || body || sizedInstruction) &&
```
Comment (tomgreenfield) [inline src/core/templates/partials/component.jsx:22]: Please wrap the template body in parentheses i.e.
```js
return (displayTitle || body || sizedInstruction) && (
  // ...
);
```

**Review by tomgreenfield — COMMENTED** [review]
Also missing changes to lockfile.

File: src/core/js/views/adaptView.js [hunk 3]
```diff
@@ -1,355 +1,397 @@
...
+    Adapt.log.deprecated(`view.childViews use view.getChildViews() and view.setChildViews([])`);
+    if (Array.isArray(this._childViews)) {
+      return _.indexBy(this._childViews, view => view.model.get('_id'));
     }
+    return this._childViews;
+  }
 
+  /**
+   * Sets an indexed by id list of child views.
+   * @deprecated since 0.5.5
```
[the remark this answers is not in the document]
Comment (oliverfoster) [reply src/core/js/views/adaptView.js:386]: 5.7.0

File: package.json [hunk 4]
```diff
@@ -16,12 +16,16 @@
...
+        "@babel/plugin-transform-react-jsx": "^7.10.4",
         "@babel/preset-env": "^7.8.4",
+        "@babel/preset-react": "^7.10.4",
         "@rollup/plugin-babel": "^5.0.4",
         "@types/backbone": "^1.4.1",
         "@types/jquery": "^3.3.31",
         "async": "^3.1.1",
         "babel-plugin-transform-amd-to-es6": "^0.3.0",
+        "babel-plugin-transform-react-templates": "^0.1.0",
+        "html-react-parser": "^0.13.0",
```
[the remark this answers is not in the document]
Comment (oliverfoster) [reply package.json:28]: Sure

File: src/core/templates/partials/component.jsx (hunk 1 above)
Comment (oliverfoster) [reply src/core/templates/partials/component.jsx:33]: Styling is currently reliant on the whitespace, the line collapses in an odd way without it.

File: src/core/templates/partials/component.jsx (hunk 2 above)
Comment (oliverfoster) [reply src/core/templates/partials/component.jsx:22]: Done

File: grunt/tasks/javascript.js [hunk 5]
```diff
@@ -157,6 +157,16 @@ module.exports = function(grunt) {
       });
     }
 
+    // Collect react templates
+    const reactTemplatePaths = [];
+    options.reactTemplates.forEach(pattern => {
+      grunt.file.expand({
+        filter: options.pluginsFilter
+      }, pattern).forEach(function(templatePath) {
```
[the remark this answers is not in the document]
Comment (oliverfoster) [reply grunt/tasks/javascript.js:165]: Yup

File: src/core/templates/partials/component.jsx (hunk 2 above)
Comment (tomgreenfield) [reply src/core/templates/partials/component.jsx]: Misssing the semi-colon.

File: src/core/templates/partials/component.jsx (hunk 1 above)
Comment (tomgreenfield) [reply src/core/templates/partials/component.jsx:33]: Okay, let's not rely on the quirkiness of yesteryear. Please could you remove the newline characters from the JSX and provide a level playing field by changing this:
https://github.com/adaptlearning/adapt_framework/blob/befc898ba0e17b58f81cc9ce7aaaf8c770864a87/src/core/templates/partials/component.hbs#L16
to:
```hbs
{{~{compile displayTitle}~}}
```
If a slight margin needs to be added then so be it (have discussed this briefly with @guywillis).

Comment (oliverfoster) [reply src/core/templates/partials/component.jsx:33]: Body and instruction also?

**Review by eleanor-heath — APPROVED** [review]

File: grunt/config/javascript.js [hunk 6]
```diff
@@ -19,6 +19,17 @@ module.exports = function(grunt, options) {
...
+          '../node_modules/react/umd/react.development.js',
+          '../node_modules/react-dom/umd/react-dom.development.js',
+          '../node_modules/html-react-parser/dist/html-react-parser.min.js'
+        ],
+        reactTemplates: [
+          '<%= sourcedir %>core/templates/**/*.jsx',
+          '<%= sourcedir %>components/*/templates/**/*.jsx',
+          '<%= sourcedir %>extensions/*/templates/**/*.jsx',
+          '<%= sourcedir %>menu/*/templates/**/*.jsx',
+          '<%= sourcedir %>theme/*/templates/**/*.jsx'
```
Comment (cahirodoherty-learningpool) [inline grunt/config/javascript.js:32]: Could these all be replaced with one `<%= sourcedir %>*/templates/**/*.jsx`?

Comment (oliverfoster) [reply grunt/config/javascript.js:32]: No, they're done like that to make them override in load order.

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

**Review by tomgreenfield — APPROVED** [review]
Magnificent.

**Comment by moloko** [comment]
Can we hold off merging this until next week's FW meeting - just to give us a chance to talk about when & how to release this?

**Comment by tomgreenfield** [comment]
While this is on hold...

### Feature request

Have the view redo its runtime properties/functions (i.e. `attributes`, `className`, `id`, `tagName`) when `this.changed()` is called.

**Review by moloko — APPROVED** [review]
OK, shall we get this merged in and released then?

**Comment by oliverfoster** [comment]
I'd just like to test that it still works first. Give me an hour or so?

**Pull request merged**